clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1062,8 +1062,17 @@ private void close(long timeout, TimeUnit timeUnit, boolean swallowException) {
         ClientUtils.closeQuietly(partitioner, "producer partitioner", firstException);
         AppInfoParser.unregisterAppInfo(JMX_PREFIX, clientId);
         log.debug("Kafka producer has been closed");
-        if (firstException.get() != null && !swallowException)
-            throw new KafkaException("Failed to close kafka producer", firstException.get());
+        Throwable exception = firstException.get();
+        if (exception != null) {
+            if (!swallowException) {
+                if (exception instanceof InterruptedException) {
+                    throw new InterruptException((InterruptedException) exception);
+                }
+                throw new KafkaException("Failed to close kafka producer", exception);
+            } else {
+                Thread.currentThread().interrupt();
```

Comment:
We should only do this if the exception is an instance of `InterruptedException`, right?

Another thing I realized.. We're only inspecting the first exception that we caught. What if the `InterruptedException` happens after we had already caught another exception?